source/solvers/mf_navier_stokes.cc

```diff
@@ -44,1083 +44,1229 @@
-    }
+      // Create transfer operators
+      mg_computing_timer.enter_subsection("Create transfer operator");
 
-  // Create transfer operator and transfer solution to mg levels
-  mg_computing_timer.enter_subsection(
-    "Create transfer operator and execute relevant transfers");
+      this->mg_transfer_ls = std::make_shared<LSTransferType>();
 
-  this->mg_transfer_ls = std::make_shared<LSTransferType>();
+      this->mg_transfer_ls->initialize_constraints(this->mg_constrained_dofs);
+      this->mg_transfer_ls->build(this->dof_handler, partitioners);
 
-  this->mg_transfer_ls->initialize_constraints(mg_constrained_dofs);
-  this->mg_transfer_ls->build(dof_handler, partitioners);
-  this->mg_transfer_ls->interpolate_to_mg(dof_handler,
-                                          mg_solution,
-                                          present_solution);
+      mg_computing_timer.leave_subsection("Create transfer operator");
+    }
+  else if (this->simulation_parameters.linear_solver
+             .at(PhysicsID::fluid_dynamics)
+             .preconditioner ==
+           Parameters::LinearSolver::PreconditionerType::gcmg)
+    {
+      // Create triangulations
+      mg_computing_timer.enter_subsection("Create level triangulations");
+      this->coarse_grid_triangulations =
+        MGTransferGlobalCoarseningTools::create_geometric_coarsening_sequence(
+          this->dof_handler.get_triangulation());
+      mg_computing_timer.leave_subsection("Create level triangulations");
+
+      // Modify the triangulations if multigrid number of levels or minimum
+      // number of cells in level are specified
+      std::vector<std::shared_ptr<const Triangulation<dim>>> temp;
+
+      int mg_min_level =
+        this->simulation_parameters.linear_solver.at(PhysicsID::fluid_dynamics)
+          .mg_min_level;
 
-  if (is_bdf(simulation_control->get_assembly_method()))
-    this->mg_transfer_ls->interpolate_to_mg(
-      dof_handler,
-      mg_time_derivative_previous_solutions,
-      time_derivative_previous_solutions);
+      AssertThrow(
+        (mg_min_level + 1) <=
+          static_cast<int>(this->coarse_grid_triangulations.size()),
+        ExcMessage(
+          "The mg min level specified is higher than the finest mg level."));
 
-  // Evaluate non linear terms for all mg operators
-  for (unsigned int level = minlevel; level <= maxlevel; ++level)
-    {
-      mg_solution[level].update_ghost_values();
-      this->mg_operators[level]->evaluate_non_linear_term(mg_solution[level]);
+      int mg_level_min_cells =
+        this->simulation_parameters.linear_solver.at(PhysicsID::fluid_dynamics)
+          .mg_level_min_cells;
 
-      if (is_bdf(simulation_control->get_assembly_method()))
-        {
-          mg_time_derivative_previous_solutions[level].update_ghost_values();
-          mg_operators[level]->evaluate_time_derivative_previous_solutions(
-            mg_time_derivative_previous_solutions[level]);
+      AssertThrow(
```

Comment:
Is the best approach here to throw or would it be to just print a warning and use the max level as the min level (e.g. no GMG)

Reply:
I am not sure about this... I will think about it and then maybe we can do this in the next MG PR